### PR TITLE
foundry-list unit test

### PR DESCRIPTION
**Pull request by iFrostizz**
Add a unit test for testing the output of `foundry-list`.
This is at the moment mocking `foundry-list` by calling `APRProof.summary`, but will be rendered more robust by refactoring the output of `APRProof.summary` to return a proper data structure.

### CODE REVIEW

**Comment by iFrostizz** [comment]
Ref: https://github.com/runtimeverification/evm-semantics/issues/1916

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import pytest
+from pyk.proof import APRProof
+
+from tests.utils import REPO_ROOT
+
+if TYPE_CHECKING:
+    from pathlib import Path
+    from typing import Final
+
+TEST_DATA = REPO_ROOT / 'kevm-pyk/src/tests/unit/test-data'
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:15]: Create a file `utils.py` in `src/test/unit`. There, define `TEST_DATA_DIR` relative to that file. (See `integration/utils.py` for an example.)

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+lists = foundry_list_expected.split('\n\n')
+assert len(lists) == len(APR_PROOFS)
+
+list_outputs = {list_output.split('\n')[0].split(': ')[1]: list_output.rstrip('\n') for list_output in lists}
+
+
+@pytest.mark.parametrize(
+    'apr_proof', APR_PROOFS, ids=[str(apr_proof.relative_to(TEST_DATA)) for apr_proof in APR_PROOFS]
+)
+def test_foundry_list(apr_proof: Path) -> None:
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:40]: This test performs logic. What it should do instead is just call `foundry_list` on a folder that contains a minimal KEVM-Foundry setup (prepared somewhere under `test-data`), and compare to the expected output.

**Review by tothtamas28 — CHANGES_REQUESTED** [review]

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 1 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:15]: Much better, thank you !

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 2 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py]: Thanks, there is now a setup foundry repo in `unit/test-data/foundry-list`

**Comment by iFrostizz** [comment]
@tothtamas28  As it's a unit test, the foundry repo is there and ready for use. But as you can see it adds a lot of files.
Should we let foundry at least handle the `forge build` to install forge-std and produce the artifacts to air up a bit on space or leave as-is ?

**Comment by tothtamas28** [comment]
> As it's a unit test, the foundry repo is there and ready for use. But as you can see it adds a lot of files. 

Right, I agree, please try to minimize this. A lot of files are necessary for constructing the proofs, but for `foundry_prove` only a small portion of these files will ever be touched.

Once the test setup is minimized, we can still mock some dependencies to reduce it further.

**Comment by iFrostizz** [comment]
I don't think that I can reduce this much more, the next step would be to foundry-kompile but it may not be a unit test anymore

File: kevm-pyk/.gitignore [hunk 3]
```diff
@@ -1,3 +1,11 @@
...
 __pycache__/
 .coverage
+src/tests/unit/test-data/foundry-list/cache/
+src/tests/unit/test-data/foundry-list/lib/
+src/tests/unit/test-data/foundry-list/out/
+!src/tests/unit/test-data/foundry-list/out/apr_proofs/
+!src/tests/unit/test-data/foundry-list/out/kompiled/
+!src/tests/unit/test-data/foundry-list/out/kompiled-llvm/
+!src/tests/unit/test-data/foundry-list/out/tmp/
+!src/tests/unit/test-data/foundry-list/out/digest
```
Comment (tothtamas28) [inline kevm-pyk/.gitignore:11]:
```suggestion
```

File: kevm-pyk/src/kevm_pyk/foundry.py [hunk 4]
```diff
@@ -664,6 +664,7 @@ def foundry_to_dot(foundry_root: Path, test: str) -> None:
 def foundry_list(foundry_root: Path) -> list[str]:
     foundry = Foundry(foundry_root)
     apr_proofs_dir = foundry.out / 'apr_proofs'
+    print(apr_proofs_dir)
```
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/foundry.py:667]: Probably leftover from debugging.

File: kevm-pyk/src/tests/unit/test-data/foundry-list/out/kompiled/allRules.txt
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test-data/foundry-list/out/kompiled/allRules.txt:1]: I'm pretty sure we don't need a kompiled definition to test this functionality. If necessary we can create a simple mock instead.

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        foundry_list_expected = f.read().rstrip()
+    assert foundry_list_expected
+
+    try:
+        run_process(['forge', 'install', '--no-git', f'foundry-rs/forge-std@{FORGE_STD_REF}'], cwd=LIST_DATA_DIR)
+    except CalledProcessError:
+        pass
+
+    run_process(['forge', 'build'], cwd=LIST_DATA_DIR)
+    list_out = '\n'.join(foundry_list(LIST_DATA_DIR))
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:31]: Just add the compiled `json` files statically if they're needed.

File: kevm-pyk/src/tests/unit/test-data/foundry-list/out/kompiled/allRules.txt
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test-data/foundry-list/out/kompiled/allRules.txt:1]: Indeed looks like we don't need it

File: kevm-pyk/.gitignore (hunk 3 above)
Comment (iFrostizz) [reply kevm-pyk/.gitignore:11]: Should we remove all those ignored folders ? We probably don't need to commit cache for instance, as well as {kompiled, kompiled-llvm, tmp} as they have shown to be unnecessary

Comment (tothtamas28) [reply kevm-pyk/.gitignore]: Right, but they shouldn't be there in the local tree either.

Comment (iFrostizz) [reply kevm-pyk/.gitignore]: Totally makes sense! Though, how do we prevent other people from committing them  once they launch the unit test ?

Comment (tothtamas28) [reply kevm-pyk/.gitignore]: My point is, unit tests should not generate these files. Whatever you need to run `foundry-list` has to be provided as a minimal set of files as test data.

Comment (iFrostizz) [reply kevm-pyk/.gitignore]: Of course, Thanks!

**Review by tothtamas28 — COMMENTED** [review]
I think this can be minimized further. Let me do some experiments with the setup, then we can discuss.

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+if TYPE_CHECKING:
+    from typing import Final
+
+LIST_DATA_DIR: Final = TEST_DATA_DIR / 'foundry-list'
+LIST_OUT: Final = LIST_DATA_DIR / 'out/'
+LIST_APR_PROOF: Final = LIST_OUT / 'apr_proofs/'
+LIST_EXPECTED: Final = LIST_DATA_DIR / 'foundry-list.expected'
+
+
+def contracts() -> dict[str, Contract]:
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:21]: This logic should not be replicated here. Instead, consider mocking `Foundry`.

Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py]: Much better, thanks !

File: kevm-pyk/src/tests/unit/test-data/foundry-list/foundry.toml
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test-data/foundry-list/foundry.toml:1]: Now that the mock is set up, is this file necessary?

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+
+LIST_DATA_DIR: Final = TEST_DATA_DIR / 'foundry-list'
+LIST_OUT: Final = LIST_DATA_DIR / 'out/'
+LIST_APR_PROOF: Final = LIST_OUT / 'apr_proofs/'
+LIST_EXPECTED: Final = LIST_DATA_DIR / 'foundry-list.expected'
+
+
+class FoundryMock:
+    out = LIST_OUT
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:32]: Consider

```suggestion
class FoundryMock:
   out: Path

    def __init__(self, out: Path)
        sef.out = out
```

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            proof_dict = json.loads(Path(proof_path).read_text())
+            pid = proof_dict['id']
+            if pid.startswith(f'{contract}.{test}'):
+                return pid
+        return ''
+
+
+def test_foundry_list(mocker: MockerFixture, update_expected_output: bool) -> None:
+    foundry_mock = mocker.patch('kevm_pyk.foundry.Foundry')
+    foundry_mock.return_value = FoundryMock()
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:63]: I think this should be

```suggestion
    foundry_mock.return_value = FoundryMock
```

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+    with LIST_EXPECTED.open() as f:
+        foundry_list_expected = f.read().rstrip()
+    assert foundry_list_expected
+
+    list_out = '\n'.join(foundry_list(LIST_DATA_DIR))
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:69]: If you add `__init__` to the mock, you can add

```suggestion
    list_out = '\n'.join(foundry_list(LIST_DATA_DIR))
    
    assert mock.called_once_with(LIST_DATA_DIR)
```

This verifies the test setup to some extent.

File: kevm-pyk/src/tests/unit/test_foundry_list.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+LIST_OUT: Final = LIST_DATA_DIR / 'out/'
+LIST_APR_PROOF: Final = LIST_OUT / 'apr_proofs/'
+LIST_EXPECTED: Final = LIST_DATA_DIR / 'foundry-list.expected'
+
+
+class FoundryMock:
+    out = LIST_OUT
+
+    @cached_property
+    def contracts(self) -> dict[str, Contract]:
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_foundry_list.py:35]: It would be good to keep mock implementation to the minimum, i.e. have a very simple implementations for `contracts` and `proof_digest`. In order to achieve that try to use simple static values. You can also simplify test data (basically rename files or their content) if necessary.

For example we don't really care if `digest` is realistic, we just want to match the value in the expected output.

File: kevm-pyk/src/tests/unit/test-data/foundry-list/foundry.toml
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test-data/foundry-list/foundry.toml:1]: It is not ! Thanks

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 7 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:32]: Thank you!

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 8 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:63]: Well then considering the other comment it should now be `FoundryMock(LIST_OUT)`, right ?

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 9 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:69]: Thanks!

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 8 above)
Comment (tothtamas28) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:63]: Well, we're mocking a class with another, so I'd expect it to work with just `FoundryMock`. Can you check to make sure?

Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:63]: Oh I get it, I think that in the above review you meant

```python
class FoundryMock:
   out: Path

    def __init__(self)
        sef.out = LIST_OUT
```

Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:63]: Calling `FoundryMock` does not seem to invoke the `__init__`

Comment (tothtamas28) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:63]: Oh yeah, that makes sense. Maybe we should use `monkeypatch` then, that's what we're using `mocker` for anyways:

```python3
from pytest import MonkeyPatch

import foundry

def test(monkeypatch: MonkeyPatch):
    monkeypatch.setattr(foundry, 'Foundry', FoundryMock)
```

But the current approach is fine as well.

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 10 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:35]: Thank you so much !!

File: kevm-pyk/src/tests/unit/test_foundry_list.py (hunk 8 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_foundry_list.py:53]: Interesting, please tell me if the current implementation fits

**Comment by iFrostizz** [comment]
The kevm library is still lazily requested from the import of the `foundry_list` and so the unit test fails on CI

File: deps/k
Comment (tothtamas28) [inline deps/k:1]: Please restore these changes.

File: kevm-pyk/src/kevm_pyk/foundry.py [hunk 11]
```diff
@@ -121,6 +120,8 @@ def digest(self) -> str:
 
     @cached_property
     def llvm_dylib(self) -> Path | None:
+        from .config import Kernel
```
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/foundry.py:123]: I think you only need these local imports to make the unit test work, the code movements is unnecessary.

```suggestion
        from .kompile import Kernel
```

**Review by tothtamas28 — APPROVED** [review]

**Comment by iFrostizz** [comment]
@tothtamas28 I'd require your input on simplifying those digest.
With the new proof system, the proofs are stored as `Contract.Method:digest`.
In that test, the digest was replaced by some static data according to your review so that it would just return `Contract.Method`.
Does that make the test too brittle, or is it okay to not test the digest ?

**Comment by tothtamas28** [comment]
> With the new proof system, the proofs are stored as `Contract.Method:digest`.
In that test, the digest was replaced by some static data according to your review so that it would just return `Contract.Method`.
Does that make the test too brittle, or is it okay to not test the digest ?

It's fine both ways. If computing the digest does not introduce dependencies that are problematic, let's compute it. Otherwise, let's just stick with the mock.

**Review by tothtamas28 — APPROVED** [review]
